### PR TITLE
Blockstore: track when all primary-index data has been purged

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
#33419 changes the keys of special-column entries; it also adds a fallback so that nodes with mixed data can read both key versions, and checks the old version when the new-version key cannot be found. However, once a node has purged all of its old-format keys, there is no reason to spend these additional db read attempts

#### Summary of Changes
Track the highest_primary_index_slot in memory
Set this value to None when all the primary-index-key data has been purged
Use this value to optimize various Blockstore read methods, as well as `purge_special_columns_exact()`

Incidentally, I was originally planning to add to this PR logic to clear the signature-only TransactionMemos in the CompactionFilter when all the primary-index data has been purged. But since the filters are built before the Blockstore element is created, it's a little tricky. Will approach that in a follow-up.

### CODE REVIEW

File: ledger/src/blockstore/blockstore_purge.rs [hunk 1]
```diff
@@ -364,8 +370,14 @@ impl Blockstore {
 
         let mut index0 = self.transaction_status_index_cf.get(0)?.unwrap_or_default();
         let mut index1 = self.transaction_status_index_cf.get(1)?.unwrap_or_default();
+        let r_highest_primary_index_slot = self.highest_primary_index_slot.read().unwrap();
+        let highest_primary_index_slot = *r_highest_primary_index_slot;
+        drop(r_highest_primary_index_slot);
```
Comment (steviez) [inline ledger/src/blockstore/blockstore_purge.rs:375]:
```suggestion
        let highest_primary_index_slot = self.get_highest_primary_index_slot();
```
nit: use the existing getter

**Review by steviez — COMMENTED** [review]
Logic all looks good, just the one nit.

> Incidentally, I was originally planning to add to this PR logic to clear the signature-only TransactionMemos in the CompactionFilter when all the primary-index data has been purged. But since the filters are built before the Blockstore element is created, it's a little tricky. Will approach that in a follow-up.

Sounds good!

File: ledger/src/blockstore/blockstore_purge.rs (hunk 1 above)
Comment (CriesofCarrots) [reply ledger/src/blockstore/blockstore_purge.rs:375]: Oh thanks! I forgot to fixup after I added the getter

**Review by steviez — APPROVED** [review]
LGTM

**Pull request merged**